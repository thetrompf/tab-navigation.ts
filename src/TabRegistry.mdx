---
name: TabRegistry
menu: Util
route: /tab-registry
---

import { Playground } from 'docz';

# Tab registry

Hello, I'm a mdx file!
