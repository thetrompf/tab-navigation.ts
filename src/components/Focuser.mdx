---
name: Focuser
menu: Component
route: /focuser
---

import { Playground } from 'docz';
import { Focuser } from './Focuser';

# Focuser

<Playground>
    <Focuser />
</Playground>
