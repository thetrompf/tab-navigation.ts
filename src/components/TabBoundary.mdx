---
name: TabBoundary
menu: Component
route: /tab-boundary
---

import { Playground, PropsTable } from 'docz';
import { TabBoundary } from './TabBoundary';
import { Input } from './Tabbable';

# Tab boundary

A super awesome description of what a tab boundary is, and what problems it solves
regarding keyboard navigation in complex editors, and/or field structures.

## Basic usage

<Playground>
    <TabBoundary>
        <Input autoComplete="off" name="username" placeholder="Username" /><br />
        <Input autoComplete="off" name="password" placeholder="Password" type="password" />
    </TabBoundary>
</Playground>

## Props

