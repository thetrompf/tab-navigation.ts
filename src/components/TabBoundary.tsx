import * as React from 'react';
import { TabRegistry } from '../TabRegistry';
import { assertNeverNonThrow, filterPropKeys, UnpackedHTMLElement } from '../util';

function hasNameProperty<T>(obj: T): obj is T & { name: string } {
    return obj != null && typeof (obj as any).name === 'string';
}

export const NavigationContext = React.createContext<TabRegistry | null>(null);

interface ComponentProps<TComp extends keyof JSX.IntrinsicElements> {
    // tslint:disable-next-line:no-reserved-keywords
    as?: TComp;
    boundaryKey?: string;
    cycle?: boolean;
    focusParentOnChildOrigin?: boolean;
    focusParentOnEscape?: boolean;
}

type Props<TComp extends keyof JSX.IntrinsicElements = 'div'> = React.HTMLAttributes<
    UnpackedHTMLElement<JSX.IntrinsicElements[TComp]>
> &
    ComponentProps<TComp>;

interface State {}

type PropsWithTabRegistry<TComp extends keyof JSX.IntrinsicElements> = Props<TComp> & {
    parentRegistry: TabRegistry | null;
};

class TabBoundaryWithTabRegistry<TComp extends keyof JSX.IntrinsicElements = 'div'> extends React.Component<
    PropsWithTabRegistry<TComp>,
    State
> {
    public static displayName = 'TabBoundary';

    private tabRegistry: TabRegistry<string>;

    public constructor(props: PropsWithTabRegistry<TComp>) {
        super(props);

        this.tabRegistry = new TabRegistry({
            cycle: props.cycle,
            focusParentOnChildOrigin: props.focusParentOnChildOrigin,
        });
    }

    public componentDidMount() {
        if (this.props.boundaryKey != null && this.props.parentRegistry != null) {
            this.props.parentRegistry.add(this.props.boundaryKey, this.tabRegistry);
        }
    }

    public componentWillReceiveProps(nextProps: PropsWithTabRegistry<TComp>) {
        if (this.props.cycle !== nextProps.cycle) {
            nextProps.cycle ? this.tabRegistry.enableCycle() : this.tabRegistry.disableCycle();
        }

        if (this.props.focusParentOnChildOrigin !== nextProps.focusParentOnChildOrigin) {
            this.tabRegistry.focusParentOnChildOrigin = nextProps.focusParentOnChildOrigin === true;
        }

        if (
            this.props.boundaryKey !== nextProps.boundaryKey &&
            this.props.parentRegistry != null &&
            this.props.parentRegistry.has(this.props.boundaryKey)
        ) {
            this.props.parentRegistry.delete(this.props.boundaryKey);
        }
    }

    public componentDidUpdate(prevProps: PropsWithTabRegistry<TComp>) {
        if (
            this.props.boundaryKey !== prevProps.boundaryKey &&
            this.props.parentRegistry != null &&
            this.props.boundaryKey != null
        ) {
            this.props.parentRegistry.add(this.props.boundaryKey, this.tabRegistry);
        }
    }

    public componentWillUnmount() {
        if (this.props.boundaryKey != null && this.props.parentRegistry != null) {
            this.props.parentRegistry.delete(this.props.boundaryKey);
        }
    }

    private filterPropKeys = (propKey: (keyof ComponentProps<TComp>) | 'parentRegistry') => {
        switch (propKey) {
            case 'as':
            case 'boundaryKey':
            case 'cycle':
            case 'focusParentOnEscape':
            case 'focusParentOnChildOrigin':
            case 'parentRegistry':
                return false;
            default:
                assertNeverNonThrow(propKey);
                return true;
        }
    };

    private onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
        if (e.key === 'Tab' && hasNameProperty(e.target)) {
            e.preventDefault();
            e.stopPropagation();
            if (e.shiftKey) {
                this.tabRegistry.focusPrev(e.target.name);
            } else {
                this.tabRegistry.focusNext(e.target.name);
            }
        } else if (e.key === 'Escape' && this.props.focusParentOnEscape) {
            e.preventDefault();
            e.stopPropagation();
            this.tabRegistry.focusParent();
        }

        if (this.props.onKeyDown != null) {
            (this.props as any).onKeyDown(e);
        }
    };

    public render() {
        const props = filterPropKeys<ComponentProps<TComp>, TComp, PropsWithTabRegistry<TComp>>(
            this.props,
            this.filterPropKeys,
        );

        const comp = this.props.as == null ? 'div' : this.props.as;
        const children = React.createElement(comp, { ...props, onKeyDown: this.onKeyDown }, this.props.children);

        return <NavigationContext.Provider value={this.tabRegistry}>{children}</NavigationContext.Provider>;
    }
}

type PropsWithForwardRef<TComp extends keyof JSX.IntrinsicElements> = Props<TComp> & {
    forwardedRef?: React.Ref<TabBoundaryWithTabRegistry<TComp>>;
};
class TabBoundaryWithForwardRef<TComp extends keyof JSX.IntrinsicElements = 'div'> extends React.Component<
    PropsWithForwardRef<TComp>
> {
    public static displayName = 'TabRegistry(TabBoundary)';

    private renderChildren = (parentRegistry: TabRegistry | null) => {
        const { forwardedRef, ...props } = this.props;
        return <TabBoundaryWithTabRegistry {...props} parentRegistry={parentRegistry} ref={forwardedRef} />;
    };

    public render() {
        return <NavigationContext.Consumer children={this.renderChildren} />;
    }
}

const forwardRef = <TComp extends keyof JSX.IntrinsicElements = 'div'>() =>
    React.forwardRef<TabBoundaryWithTabRegistry<TComp>, Props<TComp>>((props, ref) => (
        <TabBoundaryWithForwardRef {...props} forwardedRef={ref} />
    ));

export type TabBoundary = TabBoundaryWithTabRegistry;
export const TabBoundary = forwardRef<keyof JSX.IntrinsicElements>();
