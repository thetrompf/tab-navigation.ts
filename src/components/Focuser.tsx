import * as React from 'react';
import { FocuserOptions, TabRegistry } from '../TabRegistry';
import { NavigationContext } from './TabBoundary';

export type ArrowKey = 'ArrowUp' | 'ArrowDown' | 'ArrowLeft' | 'ArrowRight';
export type NavigationKey = ArrowKey | 'Tab' | 'ShiftTab' | 'Escape' | 'Enter' | 'Delete' | 'Space';

interface Props<T extends number | string = string> {
    autoFocus?: boolean;
    disabled?: boolean;
    focusKey: T;
    onArrowDown?: React.KeyboardEventHandler<HTMLInputElement>;
    onArrowKeys?: (focusKey: T, arrowKey: ArrowKey) => void;
    onArrowLeft?: React.KeyboardEventHandler<HTMLInputElement>;
    onArrowRight?: React.KeyboardEventHandler<HTMLInputElement>;
    onArrowUp?: React.KeyboardEventHandler<HTMLInputElement>;
    onBlur?: React.FocusEventHandler<HTMLInputElement>;
    onDelete?: React.KeyboardEventHandler<HTMLInputElement>;
    onEnter?: React.KeyboardEventHandler<HTMLInputElement>;
    onEscape?: React.KeyboardEventHandler<HTMLInputElement>;
    onFocus?: (opts?: FocuserOptions) => void;
    onNavigationKeys?: (focusKey: T, navKey: NavigationKey) => void;
    onSpace?: React.KeyboardEventHandler<HTMLInputElement>;
}
interface State {}

const styles: React.CSSProperties = {
    border: 'none',
    display: 'inline',
    float: 'left',
    fontSize: 0,
    height: 0,
    lineHeight: 0,
    margin: 0,
    outline: 'none',
    padding: 0,
    width: 0,
};

export class Focuser<TKey extends number | string = string> extends React.Component<Props<TKey>, State> {
    private refFocuser: HTMLInputElement | null = null;
    private tabRegistry: TabRegistry<TKey> | null = null;

    public componentWillUnmount() {
        if (this.tabRegistry != null) {
            this.tabRegistry.delete(this.props.focusKey);
            this.tabRegistry = null;
        }
    }

    private onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (this.props.disabled) {
            return;
        }

        let shouldPrevent = false;

        if (e.key === 'Enter') {
            if (this.props.onEnter != null) {
                shouldPrevent = true;
                this.props.onEnter(e);
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'Enter');
            }
        } else if (e.key === ' ') {
            if (this.props.onSpace != null) {
                shouldPrevent = true;
                this.props.onSpace(e);
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'Space');
            }
        } else if (e.key === 'Escape') {
            if (this.props.onEscape != null) {
                shouldPrevent = true;
                this.props.onEscape(e);
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'Escape');
            }
        } else if (e.key === 'Delete') {
            if (this.props.onDelete != null) {
                shouldPrevent = true;
                this.props.onDelete(e);
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'Delete');
            }
        } else if (e.key === 'ArrowUp') {
            if (this.props.onArrowUp != null) {
                shouldPrevent = true;
                this.props.onArrowUp(e);
            }
            if (this.props.onArrowKeys != null) {
                shouldPrevent = true;
                this.props.onArrowKeys(this.props.focusKey, 'ArrowUp');
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'ArrowUp');
            }
        } else if (e.key === 'ArrowDown') {
            if (this.props.onArrowDown != null) {
                shouldPrevent = true;
                this.props.onArrowDown(e);
            }
            if (this.props.onArrowKeys != null) {
                shouldPrevent = true;
                this.props.onArrowKeys(this.props.focusKey, 'ArrowDown');
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'ArrowDown');
            }
        } else if (e.key === 'ArrowLeft') {
            if (this.props.onArrowLeft != null) {
                shouldPrevent = true;
                this.props.onArrowLeft(e);
            }
            if (this.props.onArrowKeys != null) {
                shouldPrevent = true;
                this.props.onArrowKeys(this.props.focusKey, 'ArrowLeft');
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'ArrowLeft');
            }
        } else if (e.key === 'ArrowRight') {
            if (this.props.onArrowRight != null) {
                shouldPrevent = true;
                this.props.onArrowRight(e);
            }
            if (this.props.onArrowKeys != null) {
                shouldPrevent = true;
                this.props.onArrowKeys(this.props.focusKey, 'ArrowRight');
            }
            if (this.props.onNavigationKeys != null) {
                shouldPrevent = true;
                this.props.onNavigationKeys(this.props.focusKey, 'ArrowRight');
            }
        } else if (e.key === 'Tab') {
            if (e.shiftKey) {
                if (this.tabRegistry != null) {
                    shouldPrevent = true;
                    this.tabRegistry.focusPrev(this.props.focusKey);
                }
                if (this.props.onNavigationKeys != null) {
                    shouldPrevent = true;
                    this.props.onNavigationKeys(this.props.focusKey, 'ShiftTab');
                }
            } else {
                if (this.tabRegistry != null) {
                    shouldPrevent = true;
                    this.tabRegistry.focusNext(this.props.focusKey);
                }
                if (this.props.onNavigationKeys != null) {
                    shouldPrevent = true;
                    this.props.onNavigationKeys(this.props.focusKey, 'Tab');
                }
            }
        } else if (!(e.shiftKey || e.altKey || e.ctrlKey || e.metaKey)) {
            shouldPrevent = true;
        }

        if (shouldPrevent) {
            e.preventDefault();
            e.stopPropagation();
        }
    };
    private setFocuserRef = (ref: HTMLInputElement | null) => {
        this.refFocuser = ref;
    };

    public focus = (opts?: FocuserOptions) => {
        if (this.props.disabled || this.tabRegistry == null || this.refFocuser == null) {
            return false;
        }
        this.refFocuser.focus();
        if (this.props.onFocus) {
            this.props.onFocus(opts);
        }
        return true;
    };

    public render() {
        return (
            <NavigationContext.Consumer>
                {tabRegistry => {
                    if (this.tabRegistry != null && this.tabRegistry !== tabRegistry) {
                        this.tabRegistry.delete(this.props.focusKey);
                    }
                    if (tabRegistry != null && this.tabRegistry !== tabRegistry) {
                        tabRegistry.add(this.props.focusKey, this.focus);
                    }
                    this.tabRegistry = tabRegistry;
                    return (
                        <input
                            autoComplete="off"
                            autoFocus={this.props.autoFocus}
                            defaultValue=""
                            disabled={this.props.disabled}
                            key="focuser"
                            name={String(this.props.focusKey)}
                            onBlur={this.props.onBlur}
                            onKeyDown={this.onKeyDown}
                            ref={this.setFocuserRef}
                            style={styles}
                            tabIndex={-1}
                        />
                    );
                }}
            </NavigationContext.Consumer>
        );
    }
}
