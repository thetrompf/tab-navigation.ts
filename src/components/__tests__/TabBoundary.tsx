import * as React from 'react';
import { cleanup, render } from 'react-testing-library';
import { TabRegistry } from '../../TabRegistry';
import { Focuser } from '../Focuser';
import { NavigationContext, TabBoundary } from '../TabBoundary';
import { escape, shiftTab, space, tab } from './__helpers__/event';

describe('TabBoundary', () => {
    afterEach(cleanup);
    test('duplicate focus keys within same boundary throws', () => {
        jest.spyOn(console, 'error').mockImplementation(() => {
            return;
        });
        expect(() =>
            render(
                <TabBoundary>
                    <Focuser focusKey="focuser" />
                    <Focuser focusKey="focuser" />
                </TabBoundary>,
            ),
        ).toThrowErrorMatchingSnapshot();
        (console as any).error.mockRestore();
    });

    test(`tab beyond last focuser in boundary, focuses sibling boundary's "next" focuser in the tab direction`, () => {
        const onFocus1 = jest.fn();
        const onFocus2 = jest.fn();
        const onFocus3 = jest.fn();
        const onFocus4 = jest.fn();

        const { container } = render(
            <TabBoundary>
                <TabBoundary boundaryKey="a">
                    <Focuser focusKey="focuser1" onFocus={onFocus1} />
                    <Focuser focusKey="focuser2" onFocus={onFocus2} />
                </TabBoundary>
                <TabBoundary boundaryKey="b">
                    <Focuser focusKey="focuser3" onFocus={onFocus3} />
                    <Focuser focusKey="focuser4" onFocus={onFocus4} />
                </TabBoundary>
            </TabBoundary>,
        );

        const focuser2 = container.querySelector('[name=focuser2]') as HTMLInputElement;
        const focuser3 = container.querySelector('[name=focuser3]') as HTMLInputElement;

        tab(focuser2);
        expect(onFocus3).toHaveBeenCalledWith({ focusOrigin: 'prev' });
        shiftTab(focuser3);
        expect(onFocus2).toHaveBeenCalledWith({ focusOrigin: 'next' });
    });

    test('focus parent on escape focuses first focuser of parent boundary', () => {
        const onFocusParent = jest.fn();
        const { container } = render(
            <TabBoundary>
                <Focuser focusKey="parent-focuser" onFocus={onFocusParent} />
                <TabBoundary boundaryKey="child-boundary" focusParentOnEscape={true}>
                    <Focuser focusKey="child-focuser1" />
                    <Focuser focusKey="child-focuser2" />
                </TabBoundary>
            </TabBoundary>,
        );

        const childFocuser2 = container.querySelector('[name=child-focuser2]') as HTMLInputElement;
        escape(childFocuser2);

        expect(onFocusParent).toHaveBeenCalledWith({ focusOrigin: 'child' });
    });

    test('boundary with cycle enabled, focus the first/last focuser when crossing boundaries', () => {
        const onFocus1 = jest.fn();
        const onFocus2 = jest.fn();
        const onFocus3 = jest.fn();

        const { container } = render(
            <TabBoundary cycle={true}>
                <Focuser focusKey="focuser1" onFocus={onFocus1} />
                <Focuser focusKey="focuser2" onFocus={onFocus2} />
                <Focuser focusKey="focuser3" onFocus={onFocus3} />
            </TabBoundary>,
        );

        const focuser1 = container.querySelector('[name=focuser1]') as HTMLInputElement;
        const focuser3 = container.querySelector('[name=focuser3]') as HTMLInputElement;

        tab(focuser3);
        expect(onFocus1).toHaveBeenCalled();

        shiftTab(focuser1);
        expect(onFocus3).toHaveBeenCalled();

        expect(onFocus2).not.toHaveBeenCalled();
    });

    test('focus parent on child origin focuses the first focuser in the parent registry', () => {
        const onFocusParent = jest.fn();
        const { container } = render(
            <TabBoundary>
                <Focuser focusKey="parent-focuser" onFocus={onFocusParent} />
                <TabBoundary boundaryKey="child-boundary" focusParentOnChildOrigin={true}>
                    <Focuser focusKey="child-focuser1" />
                    <FocusParentOnSpace focusKey="child-focuser2" />
                </TabBoundary>
            </TabBoundary>,
        );

        const childFocuser2 = container.querySelector('[name=child-focuser2]') as HTMLInputElement;
        space(childFocuser2);

        expect(onFocusParent).toHaveBeenCalledWith({ focusOrigin: 'child' });
    });

    test('duplicate keys in sibling boundary does not throw', () => {
        expect(() =>
            render(
                <div>
                    <TabBoundary boundaryKey="sibling-1">
                        <Focuser focusKey="duplicate-1" />
                        <Focuser focusKey="duplicate-2" />
                    </TabBoundary>
                    <TabBoundary boundaryKey="sibling-2">
                        <Focuser focusKey="duplicate-1" />
                        <Focuser focusKey="duplicate-2" />
                    </TabBoundary>
                </div>,
            ),
        ).not.toThrowError();
    });
});

class FocusParentOnSpace extends React.Component<{ focusKey: string; onFocus?: () => void }> {
    private tabRegistry: TabRegistry | null = null;

    public componentWillUnmount() {
        this.tabRegistry = null;
    }

    private focusParent = () => {
        if (this.tabRegistry != null) {
            this.tabRegistry.focusParent();
        }
    };

    public render() {
        return (
            <NavigationContext.Consumer>
                {tabRegistry => {
                    this.tabRegistry = tabRegistry;
                    return (
                        <Focuser
                            focusKey={this.props.focusKey}
                            onFocus={this.props.onFocus}
                            onSpace={this.focusParent}
                        />
                    );
                }}
            </NavigationContext.Consumer>
        );
    }
}
