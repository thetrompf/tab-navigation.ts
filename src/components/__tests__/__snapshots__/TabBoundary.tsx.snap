// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBoundary duplicate focus keys within same boundary throws 1`] = `"The value: focuser already exists"`;

exports[`TabBoundary duplicate focus keys within same boundary throws 2`] = `"The value focuser does not exist"`;
